jupyterhub/scopes.py

```diff
@@ -436,36 +442,79 @@ def _add_subscopes(scope_name):
     return expanded_scopes
 
 
-def expand_scopes(scopes, owner=None):
+def _expand_scopes_key(scopes, owner=None, oauth_client=None):
+    """Cache key function for expand_scopes
+
+    scopes is usually a mutable list or set,
+    which can be hashed as a frozenset
+
+    For the owner, we only care about what kind they are,
+    and their name.
+    """
+    # freeze scopes for hash
+    frozen_scopes = frozenset(scopes)
+    if owner is None:
+        owner_key = None
+    else:
+        # owner key is the type and name
+        owner_key = (type(owner).__name__, owner.name)
+    if oauth_client is None:
+        oauth_client_key = None
+    else:
+        oauth_client_key = oauth_client.identifier
+    return (frozen_scopes, owner_key, oauth_client_key)
+
+
+@lru_cache_key(_expand_scopes_key)
+def expand_scopes(scopes, owner=None, oauth_client=None):
     """Returns a set of fully expanded scopes for a collection of raw scopes
 
     Arguments:
       scopes (collection(str)): collection of raw scopes
       owner (obj, optional): orm.User or orm.Service as owner of orm.APIToken
           Used for expansion of metascopes such as `self`
           and owner-based filters such as `!user`
+      oauth_client (obj, optional): orm.OAuthClient
+          The issuing OAuth client of an API token.
 
     Returns:
       expanded scopes (set): set of all expanded scopes, with filters applied for the owner
     """
     expanded_scopes = set(chain.from_iterable(map(_expand_scope, scopes)))
 
+    filter_replacements = {
+        "user": None,
+        "service": None,
+        "server": None,
+    }
+    owner_name = None
     if isinstance(owner, orm.User):
         owner_name = owner.name
-    else:
-        owner_name = None
+        filter_replacements["user"] = f"user={owner_name}"
```

Comment:
I'm struggling to read this logic with `owner_name` as a variable set only for `owner` being an `orm.User` instance, but not otherwise.

Later there is code like `if owner_name` which doesn't really check if there is an owner name, but if there is a `owner` of instance `orm.User`, then it sets `expanded_scopes |= _expand_self_scope(owner_name)`.

I couldn't come up with a concrete suggested change. I'm thinking perhaps a rename of `owner_name` to `user_owner_name` or having an inline comment next to the `if owner_name` statement explaining what goes on a bit more.

Reply:
since it's only defined when the owner is a user, we can use `user_name`.